src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.netcoreapp.cs

```diff
@@ -24,5 +40,167 @@ public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity di
+            {
+                access = FileAccess.Read;
+            }
+            if ((rights & FileSystemRights.WriteData) != 0 || ((int)rights & GENERIC_WRITE) != 0)
+            {
+                access = access == FileAccess.Read ? FileAccess.ReadWrite : FileAccess.Write;
+            }
+            return access;
+        }
+
+        private static unsafe SafeFileHandle CreateFileOpenHandle(string fullPath, FileMode mode, FileSystemRights rights, FileShare share, FileOptions options, FileSecurity security)
+        {
+            Debug.Assert(fullPath != null);
+
+            // Must use a valid Win32 constant
+            mode = (mode == FileMode.Append) ? FileMode.OpenOrCreate : mode;
+
+            // For mitigating local elevation of privilege attack through named pipes
+            // make sure we always call CreateFile with SECURITY_ANONYMOUS so that the
+            // named pipe server can't impersonate a high privileged client security context
+            // (note that this is the effective default on CreateFile2)
+            int flagsAndAttributes = (int)options | Interop.Kernel32.SecurityOptions.SECURITY_SQOS_PRESENT | Interop.Kernel32.SecurityOptions.SECURITY_ANONYMOUS;
+
+            SafeFileHandle handle;
+
+            fixed (byte* pSecurityDescriptor = security.GetSecurityDescriptorBinaryForm())
+            {
+                Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = secAttrs = new Interop.Kernel32.SECURITY_ATTRIBUTES
+                {
+                    nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES),
+                    bInheritHandle = ((share & FileShare.Inheritable) != 0) ? Interop.BOOL.TRUE : Interop.BOOL.FALSE,
+                    lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                };
+
+                using (DisableMediaInsertionPrompt.Create())
+                {
+                    handle = Interop.Kernel32.CreateFile(fullPath, (int)rights, share, ref secAttrs, mode, flagsAndAttributes, IntPtr.Zero);
+                }
+            }
+
+            ValidateFileHandle(handle, fullPath);
+
+            return handle;
+        }
+
+        private static void ValidateFileHandle(SafeFileHandle handle, string fullPath)
+        {
+            if (handle.IsInvalid)
+            {
+                // Return a meaningful exception with the full path.
+
+                // NT5 oddity - when trying to open "C:\" as a Win32FileStream,
+                // we usually get ERROR_PATH_NOT_FOUND from the OS.  We should
+                // probably be consistent w/ every other directory.
+                int errorCode = Marshal.GetLastWin32Error();
+
+                if (errorCode == Interop.Errors.ERROR_PATH_NOT_FOUND && fullPath.Length == PathInternal.GetRootLength(fullPath))
+                {
+                    errorCode = Interop.Errors.ERROR_ACCESS_DENIED;
+                }
+
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
```

Comment:
@JeremyKuhne one of the unit tests is failing with `ERROR_SUCCESS`, which becomes an `IOException` when the directory does not exist:

```
System.IO.FileSystemAclExtensionsTests.FileInfo_Create_NotFound [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.IO.DirectoryNotFoundException)
        Actual:   typeof(System.IO.IOException): The operation completed successfully. : 'W:\I\Do\Not\Exist\file.txt'
        ---- System.IO.IOException : The operation completed successfully. : 'W:\I\Do\Not\Exist\file.txt'
        Stack Trace:
          D:\corefx\src\System.IO.FileSystem.AccessControl\src\System\IO\FileSystemAclExtensions.netcoreapp.cs(197,0): at System.IO.FileSystemAclExtensions.ValidateFileHandle(SafeFileHandle handle, String fullPath)
```

This is because [Win32Marshal.GetExceptionForWin32Error](https://github.com/dotnet/corefx/blob/df5bba114e23afc4a3621917da0a1f810be19e7a/src/Common/src/CoreLib/System/IO/Win32Marshal.cs#L57) does not explicitly handle `ERROR_SUCCESS` in the switch case, so it throws the `default`.

So my assumption is that we need to add code to handle `Marshal.GetLastWin32Error()` when its value is `ERROR_SUCCESS` and consider it a success.

Reply:
When you get `ERROR_SUCCESS` you've done something wrong. The wrong here is that we've made another interop call before checking `GetLastWin32Error()`. The call happened in the dispose for the `DisableMediaInsertionPrompt`. You always have to get errors as soon as possible after P/Invoking to avoid this sort of thing.

Reply:
Maybe we should have a `Debug.Assert(errorCode != Interop.Errors.ERROR_SUCCESS)` in there.

Reply:
This still needs fixed, the simplest thing to do is move some or all of this into the using block.